openwrt: Store Kernel Debug-Info

SmithChart opened the pull request:
This change stores a Kernel with Debug-Symbols for the current
architecture in a new output directory '<outputdir>/debug'.
This allows a developer or operator of a network to store the kernel
along with the actual images. In case of a kernel oops the debug
information can be used with the script
'scripts/decode_stacktrace.sh'  in the kernel source tree to get the
names to the symbols of the stack trace.

OpenWRT already provides the CONFIG_COLLECT_KERNEL_DEBUG -option that
creates a kernel with debug-symbols in the OpenWRT output directory.
This change enables this option and copies the generated kernel to the
gluon output directory.

We found this behavior very useful since we see a Kernel oops from time to time - especially during development and in continuous testing.
Since a gluon build already generates a lot of images adding a few 100MB of kernels does not seem like a problem.

What is your opinion on this chance? If you approve this chage I would - of course - add the missing documentation.

NeoRaider commented on scripts/copy_output.lua:
```diff
@@ -53,6 +54,11 @@ for _, image in ipairs(lib.images) do
 
+local kernel_debug_source = string.format('openwrt/bin/targets/%s/kernel-debug.tar.bz2',
+	bindir)
+local kernel_debug_dest   = env.GLUON_IMAGEDIR..'/debug/'..env.BOARD..'-'..subtarget..'-kernel-debug.tar.bz2'
```
Instead of `env.BOARD..'-'..subtarget`, using the Gluon target name - found in the variable *target* - would be more appropriate for a Gluon output file.

SmithChart replied:
Ok, I'll give that a try.

NeoRaider commented on targets/generic:
```diff
@@ -44,6 +44,8 @@ config 'CONFIG_PACKAGE_ATH_DEBUG=y'
 config '# CONFIG_KERNEL_IP_MROUTE is not set'
 config '# CONFIG_KERNEL_IPV6_MROUTE is not set'
 
+config 'CONFIG_COLLECT_KERNEL_DEBUG=y'
```
Does this have an effect on the build time?

SmithChart replied:
The kernel is compiled in `openwrt/build_dir/target-<target>/linux-<arch>/linux-<version>/vmlinux` anyway. This file gets zipped and is placed into the output directory.
Impact on the build time should be below 1s.

NeoRaider replied:
For large kernels like x86, the bzip2 takes ~18s on my machine. An increase of 10% for the incremental build time for a single device :/

NeoRaider replied:
I don't have a good idea how to proceed.
- Some kind of escape hatch for fast incremental builds during development? Might lead to lengthy rebuilds on CONFIG_COLLECT_KERNEL_DEBUG changes though, not sure
- Changing the compression method (-> OpenWrt patch; maybe make this configurable for upstream)

NeoRaider replied:
Okay, I just tested a few compressions tools (standard Debian Buster packages) with a x86-64 build:
- kernel-debug.tar: 376M
- kernel-debug.tar.gz: 101M, compressed in ~12s
- kernel-debug.tar.bz2: 91M, compressed in ~15s
- kernel-debug.tar.xz: 57M, compressed in ~101s
- kernel-debug.tar.zst: 86M, compressed in ~1s

I think we have a clear winner - I'll look into bringing zstd support into OpenWrt soon.

NeoRaider replied:
https://patchwork.ozlabs.org/project/openwrt/list/?series=176744

NeoRaider commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,40 @@
+Kernel Oops
+-----------
+
+Sometimes a running Linux Kernel detects an error during runtime that can not
```
"kernel" is usually written with a lowercase "k".

SmithChart replied:
Fixed. (I left the heading as it is.)

NeoRaider commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,40 @@
+Kernel Oops
+-----------
+
+Sometimes a running Linux Kernel detects an error during runtime that can not
```
"cannot" or "can't"

SmithChart replied:
Fixed.

NeoRaider commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,40 @@
+This usually generates a stack trace that points to the location in the code
+that caused the oops.
+
+Linux Kernels in Gluon (and OpenWRT) are stripped.
```
"OpenWrt"

SmithChart replied:
Fixed.

NeoRaider commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,40 @@
+=========
+
+
+.. _kerneloops:
```
We usually add a prefixed based on the page path to make ref labels unique, for example `_dev-debugging-kernel-oops`

SmithChart replied:
I've added the prefix as you suggested and changed the link in the getting started.

NeoRaider commented on docs/user/getting_started.rst:
```diff
@@ -113,6 +113,12 @@ symlinks, you can use the following command to resolve these links while copying
+architecture.
+These can be used for debugging and should be stored along with the images to
+allow debugging of Kernel problems on devices in the field.
+See :ref:`Debugging<kerneloops>` for more information.
```
Add a space before `<`.

SmithChart replied:
Done.

NeoRaider commented on scripts/copy_output.lua:
```diff
@@ -53,6 +54,11 @@ for _, image in ipairs(lib.images) do
 
+local kernel_debug_source = string.format('openwrt/bin/targets/%s/kernel-debug.tar.bz2',
+	bindir)
+local kernel_debug_dest   = env.GLUON_IMAGEDIR..'/debug/'..target..'-kernel-debug.tar.bz2'
```
Lining up the `=` characters doesn't make much sense when the line before is wrapped.

SmithChart replied:
You are totally right. I have removed the alignment.

NeoRaider commented on scripts/copy_output.lua:
```diff
@@ -53,6 +54,11 @@ for _, image in ipairs(lib.images) do
 
+local kernel_debug_source = string.format('openwrt/bin/targets/%s/kernel-debug.tar.bz2',
+	bindir)
+local kernel_debug_dest   = env.GLUON_IMAGEDIR..'/debug/'..target..'-kernel-debug.tar.bz2'
```
Hmm, we should prefix the destination name including the site and release (and clean up old files using a wildcard, as it is done for images and the opkg repo). It should be possible to reuse some of the opkg repo code below.

SmithChart replied:
I have updated my branch to the naming scheme you suggested.

For example for  a build with the following make call:
```$ make all GLUON_TARGET=ar71xx-generic V=s -j10 GLUON_RELEASE=v2019.1.1-ffbs-19-g03876975-202004192147-parker_beta GLUON_BRANCH=parker_beta GLUON_PRIORITY=0```

I now the following files:
```
$ ls output/images/factory/gluon-ffbs-next-v2019.1.1-ffbs-19-g03876975-202004192147-parker_beta-* | head -n1
output/images/factory/gluon-ffbs-next-v2019.1.1-ffbs-19-g03876975-202004192147-parker_beta-alfa-network-ap121.bin

$ ls output/images/debug                                                                                                    
ffbs-next-v2019.1.1-ffbs-19-g03876975-202004192147-parker_beta-ar71xx-generic-kernel-debug.tar.bz2
```

Before copying new files I remove all `-kernel-debug.tar.bz2`for the given `target`.
So I you do multiple makes for different architectures you end up with a `kernel-debug.tar.bz2` for every architecture.

Is that what you intended?

SmithChart commented:
Rebased on top of master.

belzebub40k commented:
Wouldn't it be more suitable to put this into $(GLUON_OUTPUTDIR)/debug instead of $(GLUON_IMAGEDIR)/debug as this is not a (flashable) image?

belzebub40k commented:
decode_stacktrace.sh also calls the script decodecode. I think this should also be mentioned in the docs. Apart from that this seems to work as described. Yesterday I did a build with these patches to debug #1982 and the [decoded stackstrace](https://github.com/freifunk-gluon/gluon/issues/1982#issuecomment-617590482) looks fine.

SmithChart commented:
> Wouldn't it be more suitable to put this into $(GLUON_OUTPUTDIR)/debug instead of $(GLUON_IMAGEDIR)/debug as this is not a (flashable) image?

You are totally right. I guess I used `GLUON_IMAGEDIR` because it has been used right next to my code.
I added `GLUON_OUTDIR` to the `env` and updated my PR.

SmithChart commented:
> decode_stacktrace.sh also calls the script decodecode. I think this should also be mentioned in the docs. Apart from that this seems to work as described. Yesterday I did a build with these patches to debug #1982 and the [decoded stackstrace](https://github.com/freifunk-gluon/gluon/issues/1982#issuecomment-617590482) looks fine.

OK, I can add this.
I guess it's also a good idea to mention `/sys/kernel/debug/crashlog` on ar71xx.

mweinelt commented:
Build stage fails:

```
cp: cannot create regular file '/home/jenkins/workspace/gluon-ci_PR-1971/output/debug/ci-0.6+exp20200422-x86-64-kernel-debug.tar.bz2': No such file or directory
openwrt/staging_dir/hostpkg/bin/lua: scripts/target_lib.lua:114: assertion failed!
stack traceback:
	[C]: in function 'assert'
	scripts/target_lib.lua:114: in function <scripts/target_lib.lua:112>
	(tail call): ?
	scripts/copy_output.lua:69: in main chunk
	[C]: ?
make: *** [Makefile:196: all] Error 1
```

NeoRaider commented on scripts/copy_output.lua:
```diff
@@ -53,6 +54,20 @@ for _, image in ipairs(lib.images) do
+-- copy kernel image with debug symbols
+local kernel_debug_glob = string.format('%s/debug/\0-%s-kernel-debug.tar.bz2',
+        env.GLUON_OUTPUTDIR,
+        target)
```
The `mkdir` creates `${GLUON_IMAGEDIR}/debug`, but this attempts to write to `${GLUON_OUTPUTDIR}/debug`.

NeoRaider replied:
Either use `${GLUON_IMAGEDIR}` here, or define a separate variable `${GLUON_DEBUGDIR}` in the main Makefile to make this configurable, similar to `${GLUON_PACKAGEDIR}`.

SmithChart replied:
> Build stage fails:
> 
> ```
> cp: cannot create regular file '/home/jenkins/workspace/gluon-ci_PR-1971/output/debug/ci-0.6+exp20200422-x86-64-kernel-debug.tar.bz2': No such file or directory
> openwrt/staging_dir/hostpkg/bin/lua: scripts/target_lib.lua:114: assertion failed!
> stack traceback:
> 	[C]: in function 'assert'
> 	scripts/target_lib.lua:114: in function <scripts/target_lib.lua:112>
> 	(tail call): ?
> 	scripts/copy_output.lua:69: in main chunk
> 	[C]: ?
> make: *** [Makefile:196: all] Error 1
> ```
This should not have happend. I guess my test environment wasn't clean enough.

I now have defined `${GLUON_DEBUGDIR}` in the Makefile and have exported it into the environment. And I am now using it to create the directory and to copy files there...
One can now override it like `${GLUON_IMAGEDIR}`.

SmithChart commented:
> > decode_stacktrace.sh also calls the script decodecode. I think this should also be mentioned in the docs. Apart from that this seems to work as described. Yesterday I did a build with these patches to debug #1982 and the [decoded stackstrace](https://github.com/freifunk-gluon/gluon/issues/1982#issuecomment-617590482) looks fine.
> 
> OK, I can add this.
> I guess it's also a good idea to mention `/sys/kernel/debug/crashlog` on ar71xx.

Done

belzebub40k commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,51 @@
+Some more information on how to use this tool can be found at
+`LWN <https://lwn.net/Articles/592724/>`__.
+
+Obtaining Stacktraces on ar71xx
```
AFAIK crashlog is not dedicated to ar71xx but it is disabled on some architectures/targets.

https://github.com/openwrt/openwrt/blob/openwrt-19.07/config/Config-kernel.in#L31

NeoRaider replied:
With all those exclusions, that leaves us with... mips(el), I guess? Although it seems this is additionally enabled `target/linux/generic/config-4.14`, but I have no idea if the disabled OpenWrt .config setting has precedence over this defconfig file.

SmithChart replied:
I would suggest to leave it with ar71xx for now? The hint where it may be stored is still in the docs. And if anyone wants to add other targets/architectures they are free to do so anytime.

NeoRaider replied:
At least in the kernel .config, the crashlog is enabled for all targets except for x86 - the OpenWrt Kconfig symbols only limits on which archs it is possible to override this via OpenWrt's .config. This seems like a bug to me, but to fix it, we should first determine on which archs the crashlog actually works... Also note that in OpenWrt master, crashlog support hasn't been ported to Kernel 5.4 yet.

For this doc file, I would just remove the target from the title, and write something like "On many targets, ..." in the text.

SmithChart replied:
Done. Documentation now reads:
```
Obtaining Stacktraces
.....................

On many targets stacktraces can be read from the following
location after reboot: ::

    /sys/kernel/debug/crashlog

```

NeoRaider commented on Makefile:
```diff
@@ -55,7 +56,8 @@ endef
 	GLUON_TARGETSDIR GLUON_PATCHESDIR GLUON_TMPDIR GLUON_IMAGEDIR GLUON_PACKAGEDIR \
-	GLUON_SITEDIR GLUON_RELEASE GLUON_BRANCH GLUON_LANGS GLUON_BASE_FEEDS BOARD SUBTARGET
+	GLUON_SITEDIR GLUON_RELEASE GLUON_BRANCH GLUON_LANGS GLUON_BASE_FEEDS BOARD SUBTARGET \
+	GLUON_DEBUGDIR
```
Let's move `GLUON_DEBUGDIR` next to `GLUON_PACKAGEDIR`.

SmithChart replied:
Done. It's now on the same line.

SmithChart commented:
Rebased onto current master

SmithChart commented:
As far as I can see there is one discussion open here. Does anyone have an opinion on that?

rotanid commented:
> As far as I can see there is one discussion open here. Does anyone have an opinion on that?

i see two open ones, and one got an answer after your comment. please ask the related people via @ in the corresponding discussion-threads so we know what exactly is open.
if it's done, mark it as resolved.

NeoRaider commented:
I have pushed the zstd patches to the Gluon master now - please update

SmithChart commented:
> I have pushed the zstd patches to the Gluon master now - please update

I am working on an update. With the changed build system I'll need to do some extra testing. I will let you know when I am ready.

mweinelt commented:
The netlify errors will go away as soon as you'll rebase against master. Sorry for the noise.

SmithChart commented:
> I have pushed the zstd patches to the Gluon master now - please update

@NeoRaider my branch is updated to use your zstd-compressed kernels.

mweinelt commented:
FWIW: The docs changes are visible here https://deploy-preview-1971--epic-jang-844eca.netlify.app/dev/debugging.html

NeoRaider commented on scripts/copy_output.lua:
```diff
@@ -66,6 +67,20 @@ for _, images in pairs(lib.images) do
+lib.exec {'rm', '-f', kernel_debug_glob}
+local kernel_debug_source = string.format('openwrt/bin/targets/%s/kernel-debug.tar.zst',
+        bindir)
+local kernel_debug_dest = string.format('%s/%s-%s-%s-kernel-debug.tar.zst',
```
The glob and dest patterns should have `gluon-` before the site code.

SmithChart replied:
Done.
with:
```
$ make all GLUON_TARGET=ar71xx-generic V=s -j10 GLUON_RELEASE=v2020.1-182-g93cdd052-20200603xxxx-parker_beta
```

I now get an output of 
```
$ ls output/debug
gluon-ffbs-next-v2020.1-182-g93cdd052-20200603xxxx-parker_beta-ar71xx-generic-kernel-debug.tar.zst
```

An of course it is cleand up before the next build of this target.

NeoRaider commented on docs/dev/debugging.rst:
```diff
@@ -0,0 +1,51 @@
+On the other hand this means that in a case of a stack trace the unwinder
+can only print memory locations and no further debugging information.
+
+Gluon stores a compressed kernel with debug symbols for every architecture
```
"architecture" -> "target"

SmithChart commented:
@mweinelt 
> FWIW: The docs changes are visible here https://deploy-preview-1971--epic-jang-844eca.netlify.app/dev/debugging.html
Thanks for the link. I can say: It looks exactly like my local build :)

SmithChart commented:
> Thanks, this is working nicely. I have two more nitpicks, then we're ready to go.

@NeoRaider: I've done both requested changes. Please re-review.
